File: iec61131lang/src/main/java/edu/kit/iti/formal/automation/visitors/DefaultVisitor.java
package edu.kit.iti.formal.automation.visitors;

/*-
 * #%L
 * iec61131lang
 * %%
 * Copyright (C) 2016 Alexander Weigl
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import edu.kit.iti.formal.automation.scope.LocalScope;
import edu.kit.iti.formal.automation.datatypes.Any;
import edu.kit.iti.formal.automation.st.ast.*;

/**
 * Created by weigl on 21.06.14.
 *
 * @author weigl
 * @version $Id: $Id
 */
public class DefaultVisitor<T> implements Visitor<T> {

    /**
     * <p>defaultVisit.</p>
     *
     * @param visitable a {@link edu.kit.iti.formal.automation.visitors.Visitable} object.
     * @return a T object.
     */
    public T defaultVisit(Visitable visitable) {
        //return visitable.accept(this);
        return null;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(Location location) {
        return defaultVisit(location);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ArrayInitialization initializations) {
        return defaultVisit(initializations);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ArrayTypeDeclaration arrayTypeDeclaration) {
        return defaultVisit(arrayTypeDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(AssignmentStatement assignmentStatement) {
        return defaultVisit(assignmentStatement);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ExitStatement exitStatement) {
        return defaultVisit(exitStatement);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(CaseCondition.Range range) {
        return defaultVisit(range);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(CaseCondition.IntegerCondition integerCondition) {
        return defaultVisit(integerCondition);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(CaseCondition.Enumeration enumeration) {
        return defaultVisit(enumeration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(BinaryExpression binaryExpression) {
        return defaultVisit(binaryExpression);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ConfigurationDeclaration configurationDeclaration) {
        return defaultVisit(configurationDeclaration);
    }


    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(EnumerationTypeDeclaration enumerationTypeDeclaration) {
        return defaultVisit(enumerationTypeDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(RepeatStatement repeatStatement) {
        return defaultVisit(repeatStatement);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(WhileStatement whileStatement) {
        return defaultVisit(whileStatement);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(UnaryExpression unaryExpression) {
        return defaultVisit(unaryExpression);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(TypeDeclarations typeDeclarations) {
        return defaultVisit(typeDeclarations);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(CaseStatement caseStatement) {
        return defaultVisit(caseStatement);
    }


    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(StatementList statements) {
        return defaultVisit(statements);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ProgramDeclaration programDeclaration) {
        return defaultVisit(programDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ExpressionList expressions) {
        return defaultVisit(expressions);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(FunctionDeclaration functionDeclaration) {
        return defaultVisit(functionDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(FunctionCall functionCall) {
        return defaultVisit(functionCall);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ForStatement forStatement) {
        return defaultVisit(forStatement);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ResourceDeclaration resourceDeclaration) {
        return defaultVisit(resourceDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(FunctionBlockDeclaration functionBlockDeclaration) {
        return defaultVisit(functionBlockDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(ReturnStatement returnStatement) {
        return defaultVisit(returnStatement);
    }


    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(IfStatement ifStatement) {
        return defaultVisit(ifStatement);
    }

    /**
     * {@inheritDoc}
     */
    public T visit(GuardedStatement guardedStatement) {
        return defaultVisit(guardedStatement);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(FunctionBlockCallStatement fbc) {
        return defaultVisit(fbc);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(CaseStatement.Case aCase) {
        return defaultVisit(aCase);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(StringTypeDeclaration stringTypeDeclaration) {
        return defaultVisit(stringTypeDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(StructureTypeDeclaration structureTypeDeclaration) {
        return defaultVisit(structureTypeDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(SubRangeTypeDeclaration subRangeTypeDeclaration) {
        return defaultVisit(subRangeTypeDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(SimpleTypeDeclaration simpleTypeDeclaration) {
        return defaultVisit(simpleTypeDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(LocalScope localScope) {
        return defaultVisit(localScope);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(VariableDeclaration variableDeclaration) {
        return defaultVisit(variableDeclaration);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(CommentStatement commentStatement) {
        return defaultVisit(commentStatement);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(StructureInitialization structureInitialization) {
        return defaultVisit(structureInitialization);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(Deref deref) {
        return defaultVisit(deref);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(SymbolicReference symbolicReference) {
        return defaultVisit(symbolicReference);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(PointerTypeDeclaration ptd) {
        return defaultVisit(ptd);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public T visit(IdentifierInitializer init) {
        return defaultVisit(init);
    }

    @Override
    public T visit(InterfaceDeclaration interfaceDeclaration) {
        return defaultVisit(interfaceDeclaration);
    }

    @Override
    public T visit(ClassDeclaration clazz) {
        return defaultVisit(clazz);
    }

    @Override
    public T visit(MethodDeclaration method) {
        return defaultVisit(method);
    }

    @Override
    public T visit(Literal literal) {
        return defaultVisit(literal);
    }

    @Override
    public T visit(FunctionBlockCallStatement.Parameter parameter) {
        return null;
    }
}

File: iec61131lang/src/main/kotlin/edu/kit/iti/formal/automation/visitors/DefaultVisitor.kt
package edu.kit.iti.formal.automation.visitors

/*-
 * #%L
 * iec61131lang
 * %%
 * Copyright (C) 2016 Alexander Weigl
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import edu.kit.iti.formal.automation.datatypes.values.ReferenceValue
import edu.kit.iti.formal.automation.scope.Scope
import edu.kit.iti.formal.automation.st.ast.*

/**
 * Created by weigl on 21.06.14.
 *
 * @author weigl
 * @version $Id: $Id
 */
open class DefaultVisitor<T>() : DefaultVisitorNN<T?>() {
    override fun defaultVisit(obj: Any?): T? = null
}

abstract class DefaultVisitorNN<T> : Visitor<T> {
    abstract fun defaultVisit(obj: Any?): T

    override fun visit(elements: TopLevelElements): T = defaultVisit(elements)
    override fun visit(location: Location): T = defaultVisit(location)
    override fun visit(initializations: ArrayInitialization): T = defaultVisit(initializations)
    override fun visit(arrayTypeDeclaration: ArrayTypeDeclaration): T = defaultVisit(arrayTypeDeclaration)
    override fun visit(assignmentStatement: AssignmentStatement): T = defaultVisit(assignmentStatement)
    override fun visit(exitStatement: ExitStatement): T = defaultVisit(exitStatement)
    override fun visit(range: CaseCondition.Range): T = defaultVisit(range)
    override fun visit(integerCondition: CaseCondition.IntegerCondition): T = defaultVisit(integerCondition)
    override fun visit(enumeration: CaseCondition.Enumeration): T = defaultVisit(enumeration)
    override fun visit(binaryExpression: BinaryExpression): T = defaultVisit(binaryExpression)
    override fun visit(configurationDeclaration: ConfigurationDeclaration): T = defaultVisit(configurationDeclaration)
    override fun visit(enumerationTypeDeclaration: EnumerationTypeDeclaration): T = defaultVisit(enumerationTypeDeclaration)
    override fun visit(repeatStatement: RepeatStatement): T = defaultVisit(repeatStatement)
    override fun visit(whileStatement: WhileStatement): T = defaultVisit(whileStatement)
    override fun visit(unaryExpression: UnaryExpression): T = defaultVisit(unaryExpression)
    override fun visit(typeDeclarations: TypeDeclarations): T = defaultVisit(typeDeclarations)
    override fun visit(caseStatement: CaseStatement): T = defaultVisit(caseStatement)
    override fun visit(statements: StatementList): T = defaultVisit(statements)
    override fun visit(programDeclaration: ProgramDeclaration): T = defaultVisit(programDeclaration)
    override fun visit(expressions: ExpressionList): T = defaultVisit(expressions)
    override fun visit(functionDeclaration: FunctionDeclaration): T = defaultVisit(functionDeclaration)
    override fun visit(invocation: Invocation): T = defaultVisit(invocation)
    override fun visit(forStatement: ForStatement): T = defaultVisit(forStatement)
    override fun visit(resourceDeclaration: ResourceDeclaration): T = defaultVisit(resourceDeclaration)
    override fun visit(functionBlockDeclaration: FunctionBlockDeclaration): T = defaultVisit(functionBlockDeclaration)
    override fun visit(returnStatement: ReturnStatement): T = defaultVisit(returnStatement)
    override fun visit(ifStatement: IfStatement): T = defaultVisit(ifStatement)
    override fun visit(guardedStatement: GuardedStatement): T = defaultVisit(guardedStatement)
    override fun visit(aCase: Case): T = defaultVisit(aCase)
    override fun visit(stringTypeDeclaration: StringTypeDeclaration): T = defaultVisit(stringTypeDeclaration)
    override fun visit(structureTypeDeclaration: StructureTypeDeclaration): T = defaultVisit(structureTypeDeclaration)
    override fun visit(subRangeTypeDeclaration: SubRangeTypeDeclaration): T = defaultVisit(subRangeTypeDeclaration)
    override fun visit(simpleTypeDeclaration: SimpleTypeDeclaration): T = defaultVisit(simpleTypeDeclaration)
    override fun visit(localScope: Scope): T = defaultVisit(localScope)
    override fun visit(variableDeclaration: VariableDeclaration): T = defaultVisit(variableDeclaration)
    override fun visit(commentStatement: CommentStatement): T = defaultVisit(commentStatement)
    override fun visit(structureInitialization: StructureInitialization): T = defaultVisit(structureInitialization)
    override fun visit(deref: Deref): T = defaultVisit(deref)
    override fun visit(symbolicReference: SymbolicReference): T = defaultVisit(symbolicReference)
    override fun visit(ptd: PointerTypeDeclaration): T = defaultVisit(ptd)
    override fun visit(init: IdentifierInitializer): T = defaultVisit(init)
    override fun visit(interfaceDeclaration: InterfaceDeclaration): T = defaultVisit(interfaceDeclaration)
    override fun visit(clazz: ClassDeclaration): T = defaultVisit(clazz)
    override fun visit(method: MethodDeclaration): T = defaultVisit(method)
    override fun visit(literal: Literal): T = defaultVisit(literal)
    override fun visit(sfcStep: SFCStep): T = defaultVisit(sfcStep)
    override fun visit(actionDeclaration: ActionDeclaration): T = defaultVisit(actionDeclaration)
    override fun visit(sfcNetwork: SFCNetwork): T = defaultVisit(sfcNetwork)
    override fun visit(sfc: SFCImplementation): T = defaultVisit(sfc)
    override fun visit(transition: SFCTransition): T = defaultVisit(transition)
    override fun visit(invocation: InvocationStatement): T = defaultVisit(invocation)
    override fun visit(parameter: InvocationParameter): T = defaultVisit(parameter)
    override fun visit(referenceSpecification: ReferenceTypeDeclaration): T = defaultVisit(referenceSpecification)
    override fun visit(referenceValue: ReferenceValue): T = defaultVisit(referenceValue)
    override fun visit(globalVariableListDeclaration: GlobalVariableListDeclaration): T = defaultVisit(globalVariableListDeclaration)
    override fun visit(empty: EMPTY_EXPRESSION) = defaultVisit(empty)
}